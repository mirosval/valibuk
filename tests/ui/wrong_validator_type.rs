use valibuk::Validated;

// Define a validator using String as error type
fn is_positive(i: i32) -> Result<i32, String> {
    if i > 0 {
        Ok(i)
    } else {
        Err("wrong".to_string())
    }
}

// Define a custom error type
#[derive(Debug)]
struct E;

#[derive(Validated)]
#[validation_error(E)] // Derive using this custom error type
struct A {
    #[validator(is_positive)] // Use the String validator
    a: i32,
}

fn main() {
    let a: i32 = 1;
    // This should fail, because validator error type does not match the defined custom error type
    let instance = A::from_unvalidated(UnvalidatedA { a }).expect("valid instance");
    assert_eq!(instance.a, a);
}
