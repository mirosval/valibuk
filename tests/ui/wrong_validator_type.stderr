error[E0308]: mismatched types
  --> tests/ui/wrong_validator_type.rs:19:17
   |
16 | #[derive(Validated)]
   |          --------- expected due to this
...
19 |     #[validator(is_positive)] // Use the String validator
   |                 ^^^^^^^^^^^ expected struct `E`, found struct `String`
   |
   = note: expected fn pointer `fn(_) -> Result<_, E>`
                 found fn item `fn(_) -> Result<_, String> {is_positive}`

error[E0308]: mismatched types
  --> tests/ui/wrong_validator_type.rs:16:10
   |
16 | #[derive(Validated)]
   |          ^^^^^^^^^
   |          |
   |          expected struct `E`, found struct `String`
   |          arguments to this function are incorrect
   |
note: associated function defined here
  --> $RUST/alloc/src/vec/mod.rs
   |
   |     pub fn push(&mut self, value: T) {
   |            ^^^^
   = note: this error originates in the derive macro `Validated` (in Nightly builds, run with -Z macro-backtrace for more info)
